### PR TITLE
[WIP] Dynamic Tensor Rematerialization(Checkpointing)

### PR DESCRIPTION
**Pull request by MarisaKirisame**
We had been designing a checkpointing algorithm, Dynamic Tensor Rematerialization, that work for dynamic graph.
With our work, we had successfully save 40%-80% memory with less then double the computation cost.
WN~U@F4~%KJ@XNJR2M698`L
We want to contribute to pytorch by upstreaming the algorithm, however, we are not pytorch developers and this is our first pytorch pr. Can anyone help and discuss what should be done?
[Dynamic Tensor Rematerialization](https://arxiv.org/pdf/2006.09616.pdf) is the paper and https://github.com/uwsampl/dtr-prototype is the artifact evaluation.

### CODE REVIEW

**Comment by ezyang** [comment]
(Haven't forgotten about this, but need time to digest)

**Comment by MarisaKirisame** [comment]
We are working on other things so we arent blocked on this (yet). Let us know if there is anything we can do to help!

**Comment by ezyang** [comment]
Hi @MarisaKirisame, this is a really interesting PR. The biggest current problem with the PR is that the recording of replay has all been written out by hand; this will be a hard sell from a maintenance perspective. I think what's realistic is for us to wait until full boxing/unboxing support, and then reimplement the manual codegen part as a boxed fallback. From there, the question is in-tree versus out-of-tree; there are very few other "invasive" changes and so I think this could fairly easily live out of tree. I can help you guys set this up, when boxed fallback is running. If the extension gets quite popular we can then talk about inclusion in core.

If you want to take a crack at boxed fallback today (which will actually be a decent thing to try, since we do have a decent amount of op coverage), the best code to look at are these:

* `./aten/src/ATen/core/dispatch/backend_fallback_test.cpp` is a simple implementation of fallback, that doesn't do anything but forward on to the next function. After you setup the checkpoint information, you're going to do this, so this will give you some information about how to set that up
* https://github.com/pytorch/pytorch/pull/41943 which adds boxed fallback for vmap; basically, it looks at the IValues, finds tensors, and then manually loops over them. It should give you some idea about what kinds of idioms you can do with IValues
* For some IValue manipulation code that looks similar to what you would have to do, check out torch/csrc/jit/runtime/register_c10_ops.cpp . This implements tracing on IValues; your rematerialization traces are basically a version of this.

**Comment by MarisaKirisame** [comment]
@ezyang is there any way to have it out of tree while not requiring user to recompile pytorch from scratch? compiling pytorch take lots of time and is quite hard to setup for a typical ML researcher that only use python.

**Comment by ezyang** [comment]
Yeah, out of tree would be a C++ extension like torchvision, no need to compile PyTorch from scratch, you can just build based off of the release version of PyTorch.

**Comment by MarisaKirisame** [comment]
@ezyang for the dtr to work we must hook all inplace operation. how can i do that with fallback?

**Comment by ezyang** [comment]
Can you tell me more about what kind of hooking you need to do? This may be hard to do.

**Comment by MarisaKirisame** [comment]
@ezyang the current way DTR handle mutation rn, is that it pretend all Tensor is a Ref<ImmutableTensor>, all inplace operation as clone and inplace on the cloned result, and DTR only work on ImmutableTensor. So I have to hook all mutable call, and figure out what argument is getting changed, and how is it related to the output.

**Comment by ezyang** [comment]
OK. With DTR we should be able to interpose in on calls to inplace (in the same way we interpose on functional calls to record the trace). So you would just need a way to get information about (1) is the op inplace and (2) the correspondence of inputs to outputs. This information can all be derived from the function schema. So yes, in principle it is doable. Not sure how easy it is in practice right now.

**Comment by MarisaKirisame** [comment]
So I can get the schema? I think that's what I need - most inplace only modify a single thing and it is easy to figure out. I will see if I hit any problem.

**Comment by ezyang** [comment]
Yup. In backend fallback you'll get an OperatorHandle which you can use to retrieve the schema. I think we even have a few examples of doing this already.

**Comment by shenh10** [comment]
It seems that the above PR is not successfully compiled. Is there a fix for it?

**Comment by shenh10** [comment]
Btw, this implementation adapts global variables to record memory usage, which only supports a single device and might cause a problem for distributed case.

**Comment by MarisaKirisame** [comment]
Hi @shenh10, I’ve chatted with @nunoplopes and @ezyang about this and long term it would be nice to get this feature into PyTorch but it looks like the best path forward involves some refactoring to related systems first.

Happy to chat more if you’re interested, and while there’s no roadmap right now for getting DTR into PyTorch, I think there’s interest in moving in that direction.

And yes, the code is garbage, we are just trying to make it work for our evaluation, and thus we do not handle multiple device.

**Comment by shenh10** [comment]
@MarisaKirisame Sure, it's my pleasure to have a deeper chat. Except for checkpointing, it would be nice to have some automatic memory-saving choices in PyTorch ecosystem. The current codebase is indeed far from ready for a formal integration and release, cause I have faced some segfaults during testing your codes. Btw, I would like to help if you need more people engaged in.

**Comment by MarisaKirisame** [comment]
@shenh10 sure, email me to figure out a time to chat!

**Comment by Jack47** [comment]
Hi @ezyang , I'm developer from SenseTime, we really needs this MR to allow pytorch training bigger models. we have spent almost half a year to move it forward on pytorch 1.11, we found some pains when developing dtr on pytorch, may you guys give some advice?

1. compiling  dtr implementation with pytorch is hard, it greatly decrease developing efficiency, is there any chance we can developing this feature as a pytorch plugin to build and test separately?
2. is there more documents and scripts to help developing pytorch? such as ut and ci to help find errors quickly
3. we want to using backend fallback to avoid wraping every operations in pytorch. Is there any news on performance of backend fallback? Or is there any benchmarks I can test in my env

**Comment by shenh10** [comment]
>  @Jack47 , did you test the MR and run a model with this feature successfully? I came across certain segfaults before and no longer look into it anymore. It's grateful if there are more people engaged-in to fix this feature. It seems the author stops to maintain it already. Any experience could share?

**Comment by Jack47** [comment]
> shenh10
This MR is DTR v2 which uses backend fallback, I'm currently working on [DTR v1](https://github.com/pytorch/pytorch/issues/62448), trying to decrease DTR overhead and make it more efficiency.

**Comment by albanD** [comment]
Hey @Jack47 

1. Given that this PR is a very early draft, there is no plan to package it in a specific way no. One thing that can be done though is to add the `ciflow/binaries` label to the PR so that the CI binary generation runs and you will be able to manually download wheels from github.
2. There are a lot of details in https://github.com/pytorch/pytorch/blob/master/CONTRIBUTING.md for CI, you can simply send a PR (leave it as draft if not ready for review) and you will get CI signal from there.
3. No specific news that I know of. But for any workload that is not framework-overhead bound, it should be pretty minimal. I am not aware of any benchmark on this but things like Conjugate View are implemented with that.
4. Not sure what you mean by that? Creating new ops from c++ is rather simple via the `TORCH_LIBRARY` macro and there is now a python binding for it at `torch.library`.

I also wanted to mention that if you're seriously interested in reviving this PR, we might want to sit down and discuss the overall design of this PR. I think there are ways to do this in a much simpler way (with higher overhead) that might be enough for experimentation and/or non-overhead dominated models.

**Comment by Jack47** [comment]
> Hey @Jack47
> 
> 1. Given that this PR is a very early draft, there is no plan to package it in a specific way no. One thing that can be done though is to add the `ciflow/binaries` label to the PR so that the CI binary generation runs and you will be able to manually download wheels from github.
> 2. There are a lot of details in https://github.com/pytorch/pytorch/blob/master/CONTRIBUTING.md for CI, you can simply send a PR (leave it as draft if not ready for review) and you will get CI signal from there.
> 3. No specific news that I know of. But for any workload that is not framework-overhead bound, it should be pretty minimal. I am not aware of any benchmark on this but things like Conjugate View are implemented with that.
> 4. Not sure what you mean by that? Creating new ops from c++ is rather simple via the `TORCH_LIBRARY` macro and there is now a python binding for it at `torch.library`.
> 
> I also wanted to mention that if you're seriously interested in reviving this PR, we might want to sit down and discuss the overall design of this PR. I think there are ways to do this in a much simpler way (with higher overhead) that might be enough for experimentation and/or non-overhead dominated models.

Hey @albanD , thanks for quick response, those advices are useful, I'll read CONTRIBUTING.md carefully and use github ci to help me build wheel.

for 4, I mean there are existing old pytorch plugins that are using pybind11, so those plugins are not using dispatcher. 

I'm seriously interested in reviving this PR, so eager to known more about simpler way. I'm currently working on [DTR v1](https://github.com/pytorch/pytorch/issues/62448) which doesn't use backend fallback, I've spent too much time on wraping pytorch op :(.  So I want to try this PR. My goal is to use DTR in production env(A100 10 nodes) finally, so I want to find a little overhead design.

**Comment by MarisaKirisame** [comment]
> I also wanted to mention that if you're seriously interested in reviving this PR, we might want to sit down and discuss the overall design of this PR. I think there are ways to do this in a much simpler way (with higher overhead) that might be enough for experimentation and/or non-overhead dominated models.

Can I chime in on that? I aint maintaining this heavily anymore, as I moved onto other research project, but I'd love to share some knowledge/insight and get some design feedback.

**Comment by albanD** [comment]
Created a slack channel as a higher bandwidth channel to sync on this.
Anyone interested on this is welcome to join #dynamic-tensor-rematerialization on slack.
